src/activity_handlers.cpp

```diff
@@ -4494,36 +4494,33 @@ void activity_handlers::spellcasting_finish( player_activity *act, player *p )
                 p->mod_fatigue( cost );
                 break;
             case item_energy: {
-                if ( cost <= 0 ) {
+                if( cost <= 0 ) {
                     break;
                 }
                 const itype_id id = casting.energy_item();
-                const inventory& player_inv = p->inv;
-                std::vector<const item*> a_filter = player_inv.items_with( [id](const item& it ) {
+                const inventory &player_inv = p->inv;
+                std::vector<const item *> a_filter = player_inv.items_with( [id]( const item & it ) {
                     return it.typeId() == id;
-                    } );
+                } );
                 int countRemoved = 0;
-                if ( !a_filter[0]->count_by_charges() ) {
-                    for ( const item * it : a_filter ) {
-                        if ( countRemoved < cost ) {
+                if( !item( id ).count_by_charges() ) {
```

Comment:
You can use `item::count_by_charges( id )` to avoid constructing an item.

Reply:
There is `Character::use_amount(itype_id, int)`, which can be used instead of this block.

There is also `Character::use_charges`. It can be used instead of the block below.

Reply:
It turns out my IDE literally wasn't loading the entire intellisense, because it didn't find use_charges anywhere despite me looking for that exact function.
Redoing a bunch of it now.